Mac osx terminal support

enishii opened the pull request:
I have added support for the standard Mac OSX terminal. Currently run-in-terminal is able to open the terminal, but not run the code at the same time as well. The workaround to chmod +x the file and add a shebang to the file works, but this is not the intention setup of the program.

The code now checks if the operating system is osx (darwin), and sets up the correct Applescript command to open the terminal and run the code, or open the terminal at the working directory. This only works with the standard terminal in OSX, other terminal apps like iTerm are not supported.

Adding terminal arguments for Mac is not possible. You can however write command commands in the argument (separated by semicolons) field, which will be executed before the script, e.g. "cd /some/dir;". This is not something I added by the way, I just found out when adding arguments for testing. 

Most changes to the code are only executed if you're on OSX. The standard code has been left intact and should cause no problems. I did notice however that autoquotation of file paths was not working when shebang was activated. I made a small change to the existing code in case shebang was used. There might be another way of adding the quotes here, for me this works ok. Now filenames with spaces are always loaded correctly.

I have also updated the README.md file for Mac, and corrected some general typos.

As this is my first contribution to Github I hope I did everything ok, please let me know if you have any comments or need more information. Great package by the way! 👍

enishii committed:
osx terminal start commands added

If os = darwin, the (standard) terminal will now start through a small Applescript. Setting the terminal is not required, and the setting will be ignored for osx. Terminal execution arguments are not required. The default value will be ignored, only if a value is entered it will be used.

enishii committed:
Updated Mac part

enishii committed:
Update run-in-terminal.coffee

enishii committed:
Modified autoquotation

The autoquotation of file paths did not work when the shebang was used. Modified it such that quotes will now be added on both files and directories. 
Added text in options that terminal selection for osx is not required.

enishii committed:
Updated Mac part

And fixed some typos.

pohmelie commented:
Actually a talk about switching by platform at load time, like https://github.com/blueimp/atom-open-terminal-here/blob/master/index.coffee#L46
So, you don't need to change logic.

enishii committed:
Set defaults for Mac and Win

Added code to change the default terminal and terminal-exec-arguments
based on platform. For Mac they will be empty as the terminal is
started under the surface. For windows the values from the example have
been copied.

enishii commented:
I have added the code to create defaults for Mac (empty) and windows (based on example in README). For Mac some changes are still necessary 'under the hood' though, so some of the updated logic remains.

pohmelie commented:
This is not what I expect… but I understood the problem now. You need to close quotation string and you can't do it, since you need different constructions for «open file» and «open directory».
I think (since we have solution for mac) we need to break backward compatibility for more flexible settings set. Instead of `terminal with arguments` and `terminal execution argument` we need `launch file in terminal`, `launch directory in terminal` and new interpolation argument `{launcher}`, which will represent resolved by extension command.
Also, I think we should remove `autoquotation`, since user can quote all paths himself.

By the way, I think we need `cd {working_directory}` part of command for files. Since if not, what will be the current working directory?

enishii commented:
Beside the different commands on Mac for opening and showing directory you also need to add an additional end to the command. That's why I have added cmd_mac_end in the code, only for use on mac. Since there really is not other way than opening and running a command in the terminal, I suggest to leave these commands in the code, instead of having the users to fill them in fields. One missing ' and it won't work anymore. Other platforms are much easier I think. Therefore if the current setup works ok for *nix and windows I would suggest to leave the backward compatibility intact and not create new fields, unless it has advantages on other platforms too.

For a mac the cwd is not relevant it seems. (nothing changes when I enable/disable this) What is the effect of this on other systems? I see it is used in the child_process.exec command somehow?

Also how is the F6 option supposed to work? Where do the extra arguments come from if this is used?

pohmelie commented:
> I would suggest to leave the backward compatibility intact and not create new fields, unless it has advantages on other platforms too

Actually, this changes are enhancement since brings more flexibility. Count of fields will stay as is, we remove two fields and add two fields. Update for users on win and mac is transparent, they should not change anything in settings. So, if not you, I will do this thing and get all the fame :laughing: 

> For a mac the cwd is not relevant it seems.

No, no, no. I'm talking about this scenario:
* open atom in `foo` directory
* open file in atom `foo/bar/baz/hurr.py`
* start terminal here and run

hurr.py:
``` python
with open("hurr.txt", "w") as fout:
    fout.write("durr")
```

In which place `hurr.txt` appears? My prediction is `foo/hurr.txt`, but user expect `foo/bar/baz/hurr.txt` and this can be achieved via `cd foo/bar/baz && python hurr.py`.

enishii commented:
I can give it a shot (only for the fame ;), but I'm not sure if I understand the code well enough to be successful. This is also the first time I'm programming anything in coffeescript by the way. :) If I don't succeed I'll let you know.

I solved the cwd for mac by adding "cd {working_directory};" in the terminal execution argument field. When I run /foo/bar/helloworld.py the command send to the terminal is: 
cd "/foo/bar"; python "/foo/bar/helloworld.py" 
That way the working directory is changed before running the script. && instead of ; has the same effect by the way.

pohmelie commented:
@jnelissen 
> I solved the cwd for mac by adding "cd {working_directory};"

Yes, that is exactly what I mean :+1:

enishii committed:
Updated for OS X and more

Added a {launcher} parameter and updated the program to include this
parameter.

enishii commented:
Sorry for the delay, but please have a look at the latest commit. It works great here! I have updated the README as well to reflect the latest status.
Users can now use the {launchers} field for interpolation to the selected program. I have also added {args} to the parameters list, which you can add anywhere in the command if you wish.

pohmelie commented on lib/run-in-terminal.coffee:
```diff
@@ -1,13 +1,13 @@
-    quote = (s) -> if read_option("autoquotation") then "\"#{s}\"" else s
     choose = (a, b) -> if typeof(o[b]) in ["string", "number"] then o[b] else a
-    s.replace(/{([^{}]*)}/g, (a, b) -> quote(choose(a, b)))
+    s.replace(/{([^{}]*)}/g, (a, b) -> choose(a, b))
```
I think this can be written as 
``` coffee
s.replace(/{([^{}]*)}/g, choose)
```

enishii replied:
Fixed.

pohmelie commented on lib/run-in-terminal.coffee:
```diff
@@ -196,7 +199,8 @@ class ArgumentsRequester
 
-        @line_edit_view.addEventListener("focusout", @save_and_hide)
+        # Request args does not work if below line is uncommented? (OS X)
+        # @line_edit_view.addEventListener("focusout", @save_and_hide)
```
Yes, that is because `"focusout"` false fires. This works fine before some atom version. AFAIR, with this commented you can't hide line edit popup.

enishii replied:
I uncommented the line again. Here I don't see a popup then however.

pohmelie replied:
No no, I mean if it is commented you can't hide popup with changing focus of input. Anyway, if it is uncommented it is unusable. I don't really know what to do with it. :thinking:

enishii replied:
I changed focusout to blur, seems to work fine now.

pohmelie commented on lib/run-in-terminal.coffee:
```diff
@@ -87,30 +87,32 @@ start_terminal = (start_path, args) =>
-            cmd.push(read_option("terminal_exec_arg"))
-            cmd.push(shebang)
-            cmd.push(file_path)
+          parameters.launcher = shebang
```
Don't break style of code. It is 4 space for inner block, not 2.

enishii replied:
Fixed.

pohmelie commented on lib/run-in-terminal.coffee:
```diff
@@ -233,6 +237,16 @@ class ArgumentsRequester
+    defaultLaunchdir = 'start /D {working_directory} C:\Windows\System32\cmd.exe /u /k cd "{working_directory}"'
+  else
+    defaultLaunchfile = 'your-favorite-terminal --foo --bar "{working_directory}" && {launcher} "{file_path}"'
+    defaultLaunchdir = 'your-favorite-terminal --foo --bar "{working_directory}"'
```
* 4 spaces indent
* underscore naming

enishii replied:
Fixed, thanks for the lesson ;)

pohmelie commented on lib/run-in-terminal.coffee:
```diff
@@ -233,6 +237,16 @@ class ArgumentsRequester
+    defaultLaunchfile = """osascript -e 'tell application \"Terminal\"' -e 'activate' -e 'tell application \"Terminal\" to do script \"cd \\\"{working_directory}\\\" && {launcher} \\\"{file_path}\\\" \"' -e 'end tell'"""
+    defaultLaunchdir = """osascript -e 'tell application \"Terminal\"' -e 'activate' -e 'tell application \"Terminal\" to do script \"cd \\\"{working_directory}\\\" \"' -e 'end tell'"""
+  when 'win32'
+    defaultLaunchfile = 'start /D {working_directory} C:\Windows\System32\cmd.exe /u /k cd "{working_directory}" & {launcher} "{file_path}"'
```
I think `&&` and `&` is unnecessary for windows and linux at all. Since terminals for them have "execution" flag.

enishii replied:
Actually it wasn't required for mac either, so removed those.

enishii replied:
Hmm, didn't look very well, it was indeed required for mac.

pohmelie commented on lib/run-in-terminal.coffee:
```diff
@@ -87,30 +87,32 @@ start_terminal = (start_path, args) =>
             shebang = read_shebang(file_path)
 
-        for pair in read_option("launchers").split(",").map(strip)
+        for pair in read_option("programs").split(",").map(strip)
```
Changing option name will break backward compatibility. So version should be bumped to `1.0.0`

enishii replied:
I reversed the change, however since more options have changed it may be necessary to update to 1.0.0 anyway?

pohmelie commented on lib/run-in-terminal.coffee:
```diff
@@ -356,68 +370,68 @@ module.exports =
             else
-
-                start_terminal(start_path)
+              start_terminal(start_path)
```
The only change is empty lines and indentation?

enishii replied:
Hmm, not sure what happened there, probably testing something. ;) Reversed it to the original.

pohmelie commented on lib/run-in-terminal.coffee:
```diff
@@ -356,68 +370,68 @@ module.exports =
             title: "Show commands in context menu"
             description: "Need restart (ctrl-alt-r)"
             type: "boolean"
+            order: 8
```
Great option! Did not know about options order :+1:

enishii replied:
Thanks, I just committed the revision as well.

enishii committed:
Updated for OS X and more

Added a {launcher} parameter and updated the program to include this
parameter.

enishii committed:
Updated for OS X and more

Added a {launcher} parameter and updated the program to include this
parameter.

pohmelie commented on lib/run-in-terminal.coffee:
```diff
@@ -233,6 +235,19 @@ class ArgumentsRequester
+
+    else
+        default_Launchfile = 'your-favorite-terminal --foo --bar "{working_directory}" {launcher} "{file_path}"'
+        default_Launchdir = 'your-favorite-terminal --foo --bar "{working_directory}"'
```
Ha, this should be `default_launchfile`, no capitals. AFAIK, [`snake_case`](https://en.wikipedia.org/wiki/Snake_case) is not approved for coffee, but since I'm from python-world and love coffee as «python instead of js», I prefer underscores.

enishii replied:
Learned a new word today. ;) Guess mine was called camelcase.

pohmelie commented on lib/run-in-terminal.coffee:
```diff
@@ -233,6 +235,19 @@ class ArgumentsRequester
+
+    when 'win32'
+        default_Launchfile = 'start /D {working_directory} C:\Windows\System32\cmd.exe /u /k cd "{working_directory}" {launcher} "{file_path}"'
+        default_Launchdir = 'start /D {working_directory} C:\Windows\System32\cmd.exe /u /k cd "{working_directory}"'
```
I think second `working_directory` and `cd` is useless, since `start` have `/D` argument, which deals with working directory.

enishii replied:
Ok, I wasn't able to test on windows before. Now I have, and updated the lines. Apparently the backslashes needed escaping as well.

enishii committed:
Updated for OS X and more

Added a {launcher} parameter and updated the program to include this
parameter.

pohmelie commented:
I will test this a little on win and make release soon. Great work! :clap:

pohmelie commented:
@jnelissen 
Ok, I add some stuff, check it out. I think mac os default configuration should escape `{working_directory}` and `{launcher}` too.

enishii commented:
Working fine here. Do you mean {working_directory} should be in quotes? It is already?

pohmelie commented:
Oh, my bad. `{working_directory}` is, but `{launcher}` is not. That what I mean. Launcher can be something like `/home/dart weider/.pyenv/versions/3.6.1/bin/python`. Check out my last commit, is done properly?

enishii commented:
In that case it is needed indeed. Your update is correct.

pohmelie commented:
Published. Thanks again!
It looks like this fix #10 and #16 :clap:
